jerry-core/debugger/jerry-debugger.c

```diff
@@ -0,0 +1,446 @@
+        next_byte_code_free_p = JMEM_CP_GET_NON_NULL_POINTER (jerry_debugger_byte_code_free_t,
+                                                              byte_code_free_p->next_cp);
+
+        next_byte_code_free_p->prev_cp = byte_code_free_p->prev_cp;
+      }
+
+      jmem_heap_free_block (byte_code_free_p,
+                            ((size_t) byte_code_free_p->size) << JMEM_ALIGNMENT_LOG);
+      return true;
+    }
+
+    case JERRY_DEBUGGER_UPDATE_BREAKPOINT:
+    {
+      JERRY_DEBUGGER_CHECK_PACKET_SIZE (jerry_debugger_receive_update_breakpoint_t);
+
+      JERRY_DEBUGGER_RECEIVE_BUFFER_AS (jerry_debugger_receive_update_breakpoint_t, update_breakpoint_p);
+
+      jmem_cpointer_t byte_code_cp;
+      memcpy (&byte_code_cp, update_breakpoint_p->byte_code_cp, sizeof (jmem_cpointer_t));
+      uint8_t *byte_code_p = JMEM_CP_GET_NON_NULL_POINTER (uint8_t, byte_code_cp);
+
+      uint32_t offset;
+      memcpy (&offset, update_breakpoint_p->offset, sizeof (uint32_t));
+      byte_code_p += offset;
+
+      JERRY_ASSERT (*byte_code_p == CBC_BREAKPOINT_ENABLED || *byte_code_p == CBC_BREAKPOINT_DISABLED);
+
+      *byte_code_p = update_breakpoint_p->is_set_breakpoint ? CBC_BREAKPOINT_ENABLED : CBC_BREAKPOINT_DISABLED;
+      return true;
+    }
+
+    case JERRY_DEBUGGER_STOP:
+    {
+      JERRY_DEBUGGER_CHECK_PACKET_SIZE (jerry_debugger_receive_type_t);
+
+      JERRY_CONTEXT (debugger_stop_exec) = true;
+      JERRY_CONTEXT (debugger_stop_context) = NULL;
+      *resume_exec_p = false;
+      return true;
+    }
+
+    case JERRY_DEBUGGER_CONTINUE:
+    {
+      JERRY_DEBUGGER_CHECK_PACKET_SIZE (jerry_debugger_receive_type_t);
+
+      JERRY_CONTEXT (debugger_stop_exec) = false;
+      JERRY_CONTEXT (debugger_stop_context) = NULL;
+
+      *resume_exec_p = true;
+      return true;
+    }
+
+    case JERRY_DEBUGGER_STEP:
+    {
+      JERRY_DEBUGGER_CHECK_PACKET_SIZE (jerry_debugger_receive_type_t);
+
+      JERRY_CONTEXT (debugger_stop_exec) = true;
+      JERRY_CONTEXT (debugger_stop_context) = NULL;
+      *resume_exec_p = true;
+      return true;
+    }
+
+    case JERRY_DEBUGGER_NEXT:
+    {
+      JERRY_DEBUGGER_CHECK_PACKET_SIZE (jerry_debugger_receive_type_t);
+
+      JERRY_CONTEXT (debugger_stop_exec) = true;
+      JERRY_CONTEXT (debugger_stop_context) = JERRY_CONTEXT (vm_top_context_p);
+      *resume_exec_p = true;
+      return true;
+    }
+
+    case JERRY_DEBUGGER_GET_BACKTRACE:
+    {
+      JERRY_DEBUGGER_CHECK_PACKET_SIZE (jerry_debugger_receive_get_backtrace_t);
+
+      jerry_debugger_send_backtrace (recv_buffer_p);
+      return true;
+    }
+
+    default:
+    {
+      jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Unexpected message.");
+      jerry_debugger_close_connection ();
+      return false;
+    }
+  }
+} /* jerry_debugger_process_message */
+
+#undef JERRY_DEBUGGER_CHECK_PACKET_SIZE
+
+/**
+ * Send to the client that a breakpoint is reached and wait for action.
```

Comment:
Send to the client that a breakpoint is reached and wait for action. -> Tell the client that a breakpoint has been hit and wait for further debugger commands.